### PR TITLE
Add DataChunk filter and Add min-max of invariant variables to dataChunk 

### PR DESCRIPTION
**Pull request by NLGithubWP**

### CODE REVIEW

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/CohortProcessor.java [hunk 1]
```diff
@@ -217,23 +219,55 @@ private Boolean checkMetaChunk(MetaChunkRS metaChunk) {
       String valueSchema = ft.getFilterSchema();
       MetaFieldRS valueMetaField = metaChunk.getMetaField(valueSchema);
       if (this.checkMetaField(valueMetaField, ft)) {
-        return true;
+        return false;
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/CohortProcessor.java:222]: BUG, should be true

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/CohortProcessor.java [hunk 2]
```diff
@@ -217,23 +219,55 @@ private Boolean checkMetaChunk(MetaChunkRS metaChunk) {
...
+    }
+
+    // 2. check cohort selection
+    Filter cohortFilter = this.cohortSelector.getFilter();
+    String checkedSchema = cohortFilter.getFilterSchema();
+    FieldRS dataField = chunk.getField(checkedSchema);
+    if (cohortFilter.getType().equals(FilterType.Range)) {
+      return this.checkDataRangeFiled(dataField, cohortFilter);
+    }
+
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/CohortProcessor.java:263]: Is it finished ? It doesn't cover valueSelection and ageSelection, thus the process logic maybe lead to error. 
And according to the reference of this method, the return value is exactly the opposite. (True should be false, and false should be true)

Comment (Zrealshadow) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/CohortProcessor.java:263]: And there is no check for HashType filter

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/RangeFilter.java [hunk 3]
```diff
@@ -87,7 +87,7 @@ public BitSet accept(String[] values) throws RuntimeException {
   @Override
   public boolean accept(Scope scope) throws RuntimeException {
     for (Scope u : acceptRangeList) {
-      if (u.isSubset(scope)) {
+      if (u.isSubset(scope) || u.isIntersection(scope)) {
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/RangeFilter.java:90]: Error.
Filter [10,19) and input value [9,14) are intersection.
But Filter can not accept the [9, 14)

File: cool-core/src/main/java/com/nus/cool/core/io/readstore/DataInvariantRangeFieldRS.java [hunk 4]
```diff
@@ -25,8 +29,14 @@ public class DataInvariantRangeFieldRS implements FieldRS {
    * @param userMetaField userMetaField
    * @param userDataField userDataField
    */
-  public DataInvariantRangeFieldRS(FieldType fieldType, int invariantIdx,
+  public DataInvariantRangeFieldRS(ByteBuffer buf, FieldType fieldType, int invariantIdx,
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/io/readstore/DataInvariantRangeFieldRS.java:32]: should add corresponding unit test

File: cool-core/src/main/java/com/nus/cool/core/util/writer/NativeDataWriter.java [hunk 5]
```diff
@@ -19,208 +19,216 @@
...
+  }
+
+  @Override
+  public void close() throws IOException {
+    // force close the out stream
+    if (!finished) {
+      finish();
     }
+  }
 }
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/util/writer/NativeDataWriter.java:234]: add new line

**Review by Zrealshadow — CHANGES_REQUESTED** [review]
Some logic should be think twice. These code may lead to Bug.
Using Meta-Info to speed up cohort query processing need more specific unit test. 
I doubt that current unit test can not cover the code branch in this part.
Maybe we should rename the `checkMetaField` into a more direct and understandable name,  for example `canSkipMetaField` ? 
This direct name can give a understandable hint for the returned value.

**Comment by Zrealshadow** [comment]
Also, solve the conflict to dev branch

**Comment by KimballCai** [comment]
also, please pass the code formatting check test.

**Pull request closed without merging**